frontend/src/pages/devices/List.tsx

```diff
@@ -1,246 +1,285 @@
+      warnings.push(<Badge variant="danger">Offline</Badge>)
     }
 
-    const getLastRefreshedWarnings = (entity : DeviceT) => {
-        const warnings: JSX.Element[] = []
+    return warnings
+  }
 
-        const lr = Date.parse(entity.lastRefreshedAt)
-        const recently = sub(new Date(), {minutes: 70})
+  const getLastRefreshedWarnings = (entity: DeviceT) => {
+    const warnings: JSX.Element[] = []
 
-        if (isBefore(lr, recently)) {
-            const distance = formatDistance(lr, new Date(), { addSuffix: true })
-            warnings.push(<>Last Data Sync: { distance }</>)
-        }
+    const lr = Date.parse(entity.lastRefreshedAt)
+    const recently = sub(new Date(), { minutes: 70 })
 
-        return warnings
+    if (isBefore(lr, recently)) {
+      const distance = formatDistance(lr, new Date(), { addSuffix: true })
+      warnings.push(<>Last Data Sync: {distance}</>)
     }
 
-    const getLastWentOfflineWarnings = (entity: DeviceT) => {
-        const warnings: JSX.Element[] = []
+    return warnings
+  }
 
-        const lwo = entity.lastWentOfflineAt
-        if (lwo === null) {
-            return warnings
-        }
+  const getLastWentOfflineWarnings = (entity: DeviceT) => {
+    const warnings: JSX.Element[] = []
 
-        const recently = sub(new Date(), {days: 1})
-        const lwond = Date.parse(entity.lastWentOnlineAt!)
-        const lwoffd = Date.parse(entity.lastWentOfflineAt!)
+    const lwo = entity.lastWentOfflineAt
+    if (lwo === null) {
+      return warnings
+    }
 
-        if (entity.rawDevice.status !== "ONLINE") {
-            const distance = formatDistance(lwoffd, new Date(), { addSuffix: true })
-            warnings.push(<>Went Offline: { distance }</>)
-        } else if (isAfter(lwond, recently)) {
-            const distance = formatDistance(lwond, new Date(), { addSuffix: true })
-            warnings.push(<>Went Online: { distance }</>)
-        }
+    const recently = sub(new Date(), { days: 1 })
+    const lwond = Date.parse(entity.lastWentOnlineAt!)
+    const lwoffd = Date.parse(entity.lastWentOfflineAt!)
 
-        return warnings
+    if (entity.rawDevice.status !== "ONLINE") {
+      const distance = formatDistance(lwoffd, new Date(), { addSuffix: true })
+      warnings.push(<>Went Offline: {distance}</>)
+    } else if (isAfter(lwond, recently)) {
+      const distance = formatDistance(lwond, new Date(), { addSuffix: true })
+      warnings.push(<>Went Online: {distance}</>)
     }
 
-    return render()
+    return warnings
+  }
+
+  return render()
 }
 
 const getLockResponsivenessWarnings = (entity: DeviceT) => {
-    const warnings: JSX.Element[] = []
+  const warnings: JSX.Element[] = []
 
-    if (entity.rawDevice.status !== "ONLINE") {
-        return warnings
+  if (entity.rawDevice.status !== "ONLINE") {
+    return warnings
+  }
+
+  const tooSoon = sub(new Date(), { minutes: 10 })
+  const expectedResponseInMinutes = 60
+  let goodCode = false
```

Comment:
I think we should sort the codes before we start iterating over them (so that no matter what order they come to us, we'll process them in a consistent order).

Reply:
On the backend we sort via
```go
                func(a, b int) bool {
			if mlcs[a].EndAt.Equal(mlcs[b].EndAt) {
				return mlcs[a].StartAt.After(mlcs[b].StartAt)
			}
			return mlcs[a].EndAt.After(mlcs[b].EndAt)
		},
```

Reply:
Consider something like:
* added a year ago (quickly), won't remove for another year
* tried to add a code recently but it never added, then we "completed" it

We'd want to show a warning for the completed code even though one is enabled.

Reply:
Maybe we just need to sort by the start time (the youngest code is the one we care about most).